chain/blocksync/client.go

```diff
@@ -0,0 +1,443 @@
+	incrt "github.com/filecoin-project/lotus/lib/increadtimeout"
+	"github.com/filecoin-project/lotus/lib/peermgr"
+)
+
+// Protocol client.
+// FIXME: Rename to just `Client`. Not done at the moment to avoid
+//  disrupting too much of the consumer code, should be done along
+//  https://github.com/filecoin-project/lotus/issues/2612.
+type BlockSync struct {
+	// Connection manager used to contact the server.
+	// FIXME: We should have a reduced interface here, initialized
+	//  just with our protocol ID, we shouldn't be able to open *any*
+	//  connection.
+	host host.Host
+
+	peerTracker *bsPeerTracker
+}
+
+func NewClient(
+	host host.Host,
+	pmgr peermgr.MaybePeerMgr,
+) *BlockSync {
+	return &BlockSync{
+		host:        host,
+		peerTracker: newPeerTracker(pmgr.Mgr),
+	}
+}
+
+// Main logic of the client request service. The provided `Request`
+// is sent to the `singlePeer` if one is indicated or to all available
+// ones otherwise. The response is processed and validated according
+// to the `Request` options. Either a `validatedResponse` is returned
+// (which can be safely accessed), or an `error` that may represent
+// either a response error status, a failed validation or an internal
+// error.
+//
+// This is the internal single point of entry for all external-facing
+// APIs, currently we have 3 very heterogeneous services exposed:
+// * GetBlocks:         Headers
+// * GetFullTipSet:     Headers | Messages
+// * GetChainMessages:            Messages
+// This function handles all the different combinations of the available
+// request options without disrupting external calls. In the future the
+// consumers should be forced to use a more standardized service and
+// adhere to a single API derived from this function.
+func (client *BlockSync) doRequest(
+	ctx context.Context,
+	req *Request,
+	singlePeer *peer.ID,
+) (*validatedResponse, error) {
+	// Validate request.
+	if req.Length == 0 {
+		return nil, xerrors.Errorf("invalid request of length 0")
+	}
+	if req.Length > MaxRequestLength {
+		return nil, xerrors.Errorf("request length (%d) above maximum (%d)",
+			req.Length, MaxRequestLength)
+	}
+	if req.Options == 0 {
+		return nil, xerrors.Errorf("request with no options set")
+	}
+
+	// Generate the list of peers to be queried, either the
+	// `singlePeer` indicated or all peers available (sorted
+	// by an internal peer tracker with some randomness injected).
+	var peers []peer.ID
+	if singlePeer != nil {
+		peers = []peer.ID{*singlePeer}
+	} else {
+		peers = client.getShuffledPeers()
+		if len(peers) == 0 {
+			return nil, xerrors.Errorf("no peers available")
+		}
+	}
+
+	// Try the request for each peer in the list,
+	// return on the first successful response.
+	// FIXME: Doing this serially isn't great, but fetching in parallel
+	//  may not be a good idea either. Think about this more.
+	startTime := build.Clock.Now()
+	// FIXME: Should we track time per peer instead of a global one?
```

Comment:
Global times to used to track "what is the expected time we will need to get a response if this client fails us"